autogpt_platform/backend/backend/data/graph.py

```diff
@@ -1112,55 +1094,100 @@ async def delete_graph(graph_id: str, user_id: str) -> int:
 
 
 async def validate_graph_execution_permissions(
-    graph_id: str, user_id: str, graph_version: Optional[int] = None
+    user_id: str, graph_id: str, graph_version: int, is_sub_graph: bool = False
 ) -> None:
     """
     Validate that a user has permission to execute a specific graph.
 
     This function performs comprehensive authorization checks and raises specific
     exceptions for different types of failures to enable appropriate error handling.
 
+    ## Logic
+    A user can execute a graph if any of these is true:
+    1. They own the graph and some version of it is still listed in their library
+    2. The graph is published in the marketplace and listed in their library
+    3. The graph is published in the marketplace and is being executed as a sub-agent
+
     Args:
         graph_id: The ID of the graph to check
         user_id: The ID of the user
-        graph_version: Optional specific version to check. If None (recommended),
-                      performs version-agnostic check allowing execution of any
-                      version as long as the graph is in the user's library.
-                      This is important for sub-graphs that may reference older
-                      versions no longer in the library.
+        graph_version: The version of the graph to check
+        is_sub_agent: Whether this is being executed as a sub-graph.
+            If `True`, the graph isn't required to be in the user's Library.
 
     Raises:
         GraphNotInLibraryError: If the graph is not in the user's library (deleted/archived)
         NotAuthorizedError: If the user lacks execution permissions for other reasons
     """
+    # Step 1: Check if user owns this graph
+    user_owns_graph = await get_graph_metadata(graph_id, graph_version) is not None
 
-    # Step 1: Check library membership (raises specific GraphNotInLibraryError)
+    # Step 2: Check if agent is in the library *and not deleted*
     where_clause: LibraryAgentWhereInput = {
         "userId": user_id,
         "agentGraphId": graph_id,
         "isDeleted": False,
-        "isArchived": False,
     }
 
     if graph_version is not None:
         where_clause["agentGraphVersion"] = graph_version
 
-    count = await LibraryAgent.prisma().count(where=where_clause)
-    if count == 0:
+    user_has_in_library = await LibraryAgent.prisma().count(where=where_clause) > 0
+
+    # Step 3: Apply permission logic
+    if user_owns_graph and user_has_in_library:
+        pass
```

Comment:
why don't we just return instead of pass

Reply:
because below there is a comment block which is an explicit placeholder for more checks